### PR TITLE
Add hash of a secret support

### PR DESCRIPTION
**Pull request by mn3m0nic**
### Description:

In order to de-duplicate findings and provide useful reports where only actionable leaked secrets are displayed needs to have a key field. Field "Secret" match this requirement, but when output will be unacceptably sensitive. To decrease the level of sensitivity need to use another field such as **SecretSHA** which will be just SHA256() of the Secret field.

### Checklist:

* [X] Does your PR pass tests?
* [X] Have you written new tests for your changes?
* [X] Have you lint your code locally prior to submission?

### CODE REVIEW

**Comment by mn3m0nic** [comment]
Potentially related with https://github.com/gitleaks/gitleaks/issues/1054

File: detect/detect.go [hunk 1]
```diff
@@ -272,6 +273,9 @@ func (d *Detector) detectRule(fragment Fragment, rule config.Rule) []report.Find
 			loc.endLineIndex = matchIndex[1]
 		}
 
+		// Calculate the SHA256 hash sum of secret
+		secretSHA := fmt.Sprintf("%x", sha256.Sum256([]byte(secret)))
```
Comment (rgmz) [inline detect/detect.go:277]: How do you feel about calling `strings.TrimSpaces` before hashing the secret? I've seen a few instances where the same secret is captured with various amounts of whitespace, and doing so would ensure they all produce the same hash.

Comment (mn3m0nic) [reply detect/detect.go:277]: Hi @rgmz . 
Currently, the variable Secret is trimmed by new line occurrences ('\n') and the hash in SecretSHA uses the same value from the Secret variable for consistency. If you are about to add additional trim of spaces (" ") to the Secret variable - I suggest doing it as a separate Pull request with an example of the issue.

**Comment by mn3m0nic** [comment]
I vote for option 2 (if secret empty return empty string `""`) by default and option 5 when enabled as a parameter "`--hash-files`" (or something close to it TBD)

**TL;DR:**

> Why 5

**SecretSHA** is intended to be a variable that could be used as a _Key_ to sort and receive a list of unique secret leaks.
For example, for a mid-size long enough git repository you may have 10K or even 50K of RAW findings in a Git history comment, but when we de-duplicate them using the "SecretSHA" (or using sensitive "Secret") field you may receive only 20 actual unique secrets which should be rotated as a first priority from all findings. For multiple projects scan, you may find the same token, password, or RSA key was used in different places. It should not increase the number of unique findings. You still have to rotate for example 20 real keys across 100 repositories and accept dozens of findings as a new baseline.

The requirements for SecretsSHA, as I see them, are next:

1. To explicitly and unambiguously identify a secret leaked content itself
2. To be non-sensitive enough to use it as a key value in reporting on different levels (including store reports)

You are suggesting adding a filename as a key identifier, which may be ok for criteria 1, but a filename is not unique enough even with a path for a multiple projects scan. Let's assume a scan of N repositories in the group.
For example in report1.json for project1 you may have a finding for the ".env" file which will be "`e9cbb0224c4a3d23a6019ba557e0cd568c1ad5e1582ff1e335fb7d99b7a1055d`" and for project2 where the file may be completely different. I see the  **content of a file** is equal to a "Secret" field here, not a filename.

As this will bring additional performance degradation when the scan will be performed, maybe it should be enabled as an option. (TBD)

> Why 2 ("") for files, not 1?

I hope an empty hash will tell the next program which will use an output of The Gitleaks as "there is no hash here" instead here is hash "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855" which you must add to the dictionary and use it as "This is hash, but there is no value, so this is not actually usable hash, so please ignore it".
No value, no hash - will make it more clear.

**Comment by rgmz** [comment]
I think you make a good argument for option 2. That was initially my first choice but I was on the fence about its purpose and usefulness.

Curious what @zricethezav thinks.

**Comment by zricethezav** [comment]
Hi @mn3m0nic and @rgmz,

First, thanks for opening this PR. I think it provides a useful feature as you've clearly stated. 

> Curious what @zricethezav thinks.

At first I thought hmmm maybe we should include some logic like:
``` 
hash = "from secret"
if secret == ""
then hash = "from file"
```

but I think @mn3m0nic brought up a good point on `Why 2 ("") for files, not 1?`.

Gonna check it out locally and play around with it before merging

**Comment by Devvox93** [comment]
This would solve my issues mentioned in #1059 as well. Different mechanics, but it achieves the same goal (being path-agnostic).
